### ISSUE TITLE
Richer Diagnostic Display Meta-Issue

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#41860]
This issue tracks a couple of different useful ideas for making diagnostics "richer" for display purposes. We've wanted some of these for a while, and they're partially inspired by testing tools like Jest and Mocha. We've also seen tools like Rust, Rome, and Flow prove them out to some extent.

Long-term, you could imagine them being interactive with the a richer compiler client.

# Printing Messages Across Lines

```
Type

    { x: string, y: number }

is not assignable to

    { x: string, y: string }
```

Notice no punctuation at the end. Not sure if we can generalize messages that way.

Also not sure how this plays well with

* tools that parse our output.
* diagnostic *elaborations*

# Highlighted Spans in Messages

Imagine being able to highlight individual sections inside of types to indicate where the issue occurs.

```
                              vvvvvvvvvvv
Type '{ a: string; b: number; c: boolean; }' is not assignable to type '{ a: string; b: number; }'.
                              ^^^^^^^^^^^
  Object literal may only specify known properties, and 'c' does not exist in type '{ a: string; b: number; }'.
```

# Diff Highlighting in Messages

Basically imagine red being the provided color, with green being an expected color.

```diff
Type '[string, number, boolean]' is not assignable to type '[string, string, boolean]'.
               ^^^^^^                                                ^^^^^^
               rrrrrr                                                gggggg
```

**Comment by weswigham** [comment]
I think adding extra lines of (potentially colored) text (and linebreaks) under `--pretty` style output is probably safe (since the file name/line/error code is what [vscode matches matches on](https://github.com/microsoft/vscode/blob/master/extensions/typescript-language-features/package.json#L1045), since that's the only "stableish" part of the message).

**Comment by DanielRosenwasser** [comment]
@dbaeumer may have some more knowledge on potential problem matcher issues, but I know other tools split messages across multiple lines too.

**Comment by dbaeumer** [comment]
Passing on to @alexr00 (tasks do have multi line matchers but I am not sure what the current state is)

**Comment by alexr00** [comment]
Here's the documentation on multiline problem matchers in VS Code, but I don't think you'd need them if you're only adding extra lines: https://code.visualstudio.com/Docs/editor/tasks#_defining-a-multiline-problem-matcher

Adding extra lines that don't need to read by a VS Code problem matcher should just be ignored by the existing typescript problem matcher.

**Comment by DanielRosenwasser** [comment]
Another one: smarter truncation don't elide the parts that are actually gonna be elaborated on (@RyanCavanaugh calls this "differential `typeToString` and wants to prototype it).

**Comment by DanielRosenwasser** [comment]
I opened up #42597 to track that, since it's not as relevant to richer diagnostic display.

**Comment by JoshuaKGoldberg** [comment]
@DanielRosenwasser I'd be very interested in contributing here. Are there particular parts of this that could already be split out into `Help Wanted` tasks?

**Comment by typeholes** [comment]
I have been playing with including the diagnostic arguments in the diagnostics with a type of 
```ts 
{ text: string, type: 'Node'|'Type'|'string', cacheId: number }
```
and adding a function to fetch the relevant object by the cache id.  This enables a lot of functionality for displaying and interacting with error messages.  For example, I use it to allow the user to expand a type or display the value declaration of a symbol.  The big upshot is flexibility - Diagnostic functionality is not limited to what is baked into typescript.

Is this issue a good place to discuss this approach, or should I open a new one?

**Comment by michAtEl** [comment]
When I hover over a type alias in VS Code, VS Code  shows a description of the type. If this description is long (see the following example), VS Code truncates this description. It would be great to have either 
- a configurable truncation limit
- a "Show full type" action in the hover tooltip, or
- an expandable/scrollable hover panel for large types

Example:
```
type Alpha =
    | "a" | "b" | "c" | "d" | "e" | "f" | "g" | "h" | "i" | "j"
    | "k" | "l" | "m" | "n" | "o" | "p" | "q" | "r" | "s" | "t"
    | "u" | "v" | "w" | "x" | "y" | "z";

type Pair = `${Alpha}${Alpha}`;
//   ^? -- truncated: "aa" | "ab" | "ac" | ... 672 more ...
```
